File: backend/app/schemas/user_data.py
"""Схемы для пользовательских данных: избранное, оценки, история."""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import ImageURLs


# ─── Избранное (watchlist) ───────────────────────────────────────


class FavoriteAddRequest(BaseModel):
    """Добавить в избранное."""
    status: Literal["want_to_watch", "watching", "watched", "dropped"] = "want_to_watch"
    note: str | None = Field(None, max_length=2000)


class FavoriteItem(BaseModel):
    """Элемент избранного с данными сущности."""
    entity_id: int
    entity_type: str
    title: str
    summary: str | None = None
    images: ImageURLs = ImageURLs()
    release_year: int | None = None
    status: str
    note: str | None = None
    added_at: datetime
    watched_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class FavoritesResponse(BaseModel):
    items: list[FavoriteItem]
    total: int
    limit: int
    offset: int


class FavoriteCheckResponse(BaseModel):
    """Быстрая проверка для иконки сердечка на карточке."""
    is_favorite: bool
    status: str | None = None


# ─── Пользовательские оценки ─────────────────────────────────────


class RateRequest(BaseModel):
    rating: int = Field(..., ge=1, le=10, description="Оценка от 1 до 10")
    would_recommend: bool | None = None


class RatingItem(BaseModel):
    entity_id: int
    rating: int
    would_recommend: bool | None
    rated_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EntityRatingStats(BaseModel):
    """Сводная статистика оценок сущности (от всех пользователей)."""
    avg_rating: float | None = None
    total_ratings: int = 0
    recommend_percent: float | None = None  # сколько % рекомендуют (would_recommend=true)


# ─── История ─────────────────────────────────────────────────────


class SearchHistoryItem(BaseModel):
    query_text: str
    results_count: int
    searched_at: datetime
    language_code: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ViewHistoryItem(BaseModel):
    entity_id: int
    entity_type: str
    title: str
    images: ImageURLs = ImageURLs()
    viewed_at: datetime
    duration_seconds: int | None = None

    model_config = ConfigDict(from_attributes=True)


class HistoryResponse(BaseModel):
    """Универсальный ответ для истории — поиски ИЛИ просмотры."""
    searches: list[SearchHistoryItem] = []
    views: list[ViewHistoryItem] = []

File: front/src/api/userData.ts
import { apiClient } from './client';
import type { ImageURLs } from './types';

export interface FavoriteItem {
  entity_id: number;
  entity_type: string;
  title: string;
  summary?: string | null;
  images: ImageURLs;
  release_year?: number | null;
  status: string;
  note?: string | null;
  added_at: string;
  watched_at?: string | null;
}

export interface FavoritesResponse {
  items: FavoriteItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface FavoriteCheckResponse {
  is_favorite: boolean;
  status: string | null;
}

export interface RatedFilmItem {
  entity_id: number;
  entity_type: string;
  title: string;
  images: ImageURLs;
  release_year?: number | null;
  rating: number;
  would_recommend?: boolean | null;
  rated_at: string;
  updated_at: string;
}

export interface MyRatingsResponse {
  items: RatedFilmItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface RatingBucket {
  rating: number;
  count: number;
}

export interface RatingDistribution {
  buckets: RatingBucket[];
  total: number;
  average: number | null;
}

export interface ProfileStats {
  ratings_count: number;
  favorites_count: number;
  watched_count: number;
  want_to_watch_count: number;
  views_count: number;
  searches_count: number;
}

export interface SearchHistoryItem {
  query_text: string;
  results_count: number;
  searched_at: string;
  language_code?: string | null;
}

export interface ViewHistoryItem {
  entity_id: number;
  entity_type: string;
  title: string;
  images: ImageURLs;
  viewed_at: string;
  duration_seconds?: number | null;
}

export interface HistoryResponse {
  searches: SearchHistoryItem[];
  views: ViewHistoryItem[];
}

export async function getFavorites(params?: {
  type?: 'film' | 'person';
  limit?: number;
  offset?: number;
}): Promise<FavoritesResponse> {
  const { data } = await apiClient.get<FavoritesResponse>('/api/favorites', {
    params,
  });
  return data;
}

export async function checkFavorite(entityId: number): Promise<FavoriteCheckResponse> {
  const { data } = await apiClient.get<FavoriteCheckResponse>(
    `/api/favorites/check/${entityId}`,
  );
  return data;
}

export async function addFavorite(entityId: number): Promise<void> {
  await apiClient.post(`/api/favorites/${entityId}`);
}

export async function removeFavorite(entityId: number): Promise<void> {
  await apiClient.delete(`/api/favorites/${entityId}`);
}

export async function getMyRatings(params?: {
  limit?: number;
  offset?: number;
}): Promise<MyRatingsResponse> {
  const { data } = await apiClient.get<MyRatingsResponse>('/api/ratings/me', {
    params,
  });
  return data;
}

export async function getRatingDistribution(): Promise<RatingDistribution> {
  const { data } = await apiClient.get<RatingDistribution>(
    '/api/ratings/me/distribution',
  );
  return data;
}

export async function getProfileStats(): Promise<ProfileStats> {
  const { data } = await apiClient.get<ProfileStats>('/api/profile/stats');
  return data;
}

export async function getHistory(limit = 50): Promise<HistoryResponse> {
  const { data } = await apiClient.get<HistoryResponse>('/api/history', {
    params: { limit },
  });
  return data;
}
